impl/src/main/java/org/jboss/weld/bootstrap/events/AfterBeanDiscoveryImpl.java

```diff
@@ -80,11 +85,69 @@ public void addBean(Bean<?> bean) {
         Preconditions.checkArgumentNotNull(bean, "bean");
         ExternalBeanAttributesFactory.validateBeanAttributes(bean, getBeanManager());
         validateBean(bean);
-        additionalBeans.add(bean);
+        additionalBeans.add(new BeanRegistration(bean));
     }
 
-    protected <T> void processBean(Bean<T> bean) {
-        BeanManagerImpl beanManager = getOrCreateBeanDeployment(bean.getBeanClass()).getBeanManager();
+    @Override
+    public void addContext(Context context) {
+        checkWithinObserverNotification();
+        Preconditions.checkArgumentNotNull(context, "context");
+        Class<? extends Annotation> scope = context.getScope();
+        if (scope == null) {
+            throw ContextLogger.LOG.contextHasNullScope(context);
+        }
+        if (!getBeanManager().isScope(scope)) {
+            MetadataLogger.LOG.contextGetScopeIsNotAScope(scope, context);
+        }
+        if (scope == ApplicationScoped.class || scope == Dependent.class) {
+            throw ContextLogger.LOG.cannotRegisterContext(scope, context);
+        }
+        getBeanManager().addContext(context);
+    }
+
+    @Override
+    public void addObserverMethod(ObserverMethod<?> observerMethod) {
+        checkWithinObserverNotification();
+        Preconditions.checkArgumentNotNull(observerMethod, "observerMethod");
+        validateObserverMethod(observerMethod, getBeanManager(), null);
+        additionalObservers.add(new ForwardingExperimentalObserverMethod<>(observerMethod));
+    }
+
+    @Override
+    public <T> AnnotatedType<T> getAnnotatedType(Class<T> type, String id) {
+        checkWithinObserverNotification();
+        Preconditions.checkArgumentNotNull(type, TYPE_ARGUMENT_NAME);
+        return slimAnnotatedTypeStore.get(type, id);
+    }
+
+    @Override
+    public <T> Iterable<AnnotatedType<T>> getAnnotatedTypes(Class<T> type) {
+        checkWithinObserverNotification();
+        Preconditions.checkArgumentNotNull(type, TYPE_ARGUMENT_NAME);
+        return cast(slimAnnotatedTypeStore.get(type));
+    }
+
+    @Override
+    public <T> BeanBuilder<T> addBean() {
+        Class<? extends Extension> extensionClass = getReceiver().getClass();
+        BeanManagerImpl beanManager = getOrCreateBeanDeployment(extensionClass).getBeanManager();
```

Comment:
It's fine to use extension class as the default bean class. We should however defer determining beanManager so that  if a bean class is provided explicitly, we use BM for this explicit class. Current eager approach makes that impossible.